lib/Sema/NameBinding.cpp

```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
+        // Get the arg, which should be in a paren expression.
+        auto fnNameExpr = dyn_cast<UnresolvedDeclRefExpr>(CE->getFn());
+        if (!fnNameExpr) {
+          return false;
+        }
+        auto fnName = fnNameExpr->getName().getBaseName().str();
+        auto *PE = dyn_cast<ParenExpr>(CE->getArg());
+        if (!PE) {
+          return false;
+        }
+        if (fnName.equals("_compiler_version")) {
+          auto SLE = cast<StringLiteralExpr>(PE->getSubExpr());
+
+          // Parse the compiler version string.  Diagnostics have already been
+          // emitted so suppress them.
+          DiagnosticTransaction trans(Binder.Context.Diags);
+          auto versionRequirement =
+            version::Version::parseCompilerVersionString(SLE->getValue(),
+                                                         SLE->getLoc(),
+                                                         &Binder.Context.Diags);
+          trans.abort();
+
+          auto thisVersion = version::Version::getCurrentCompilerVersion();
+          auto VersionNewEnough = thisVersion >= versionRequirement;
+          return VersionNewEnough;
+        } else if (fnName.equals("swift")) {
+          auto PUE = dyn_cast<PrefixUnaryExpr>(PE->getSubExpr());
+          if (!PUE) {
+            return false;
+          }
+          auto prefix = dyn_cast<UnresolvedDeclRefExpr>(PUE->getFn());
+          auto versionArg = PUE->getArg();
+          auto versionStartLoc = versionArg->getStartLoc();
+          auto endLoc = Lexer::getLocForEndOfToken(Context.SourceMgr,
+                                                   versionArg->getSourceRange().End);
+          CharSourceRange versionCharRange(Context.SourceMgr, versionStartLoc,
+                                           endLoc);
+          auto versionString = Context.SourceMgr.extractText(versionCharRange);
+
+          // Parse the version string.  Diagnostics have already been emitted
+          // so suppress them.
+          DiagnosticTransaction trans(Binder.Context.Diags);
+          auto versionRequirement =
+            version::Version::parseVersionString(versionString,
+                                                 versionStartLoc,
+                                                 &Binder.Context.Diags);
+          trans.abort();
+
+          assert(versionRequirement.hasValue());
+          auto thisVersion = Context.LangOpts.EffectiveLanguageVersion;
+          assert(prefix->getName().getBaseName().str().equals(">="));
+
+          auto VersionNewEnough = thisVersion >= versionRequirement.getValue();
+          return VersionNewEnough;
+        } else {
+          auto UDRE = cast<UnresolvedDeclRefExpr>(PE->getSubExpr());
+
+          // The sub expression should be an UnresolvedDeclRefExpr (we won't
+          // tolerate extra parens).
+          auto argument = UDRE->getName().getBaseName().str();
+          // FIXME: Perform the replacement macOS -> OSX in Parse if possible.
```

Comment:
Also, "finish making 'macOS' the canonical name". \*sigh\* Unfinished work from me.

(not really part of the review, just complaining out loud)